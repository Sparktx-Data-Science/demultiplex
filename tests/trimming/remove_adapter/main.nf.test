nextflow_pipeline {

    name "Trimming - Remove Adapter"
    script "../../../main.nf"
    profile "test_bclconvert"
    tag "remove_adapter"
    tag "pipeline"

    test("remove_adapter = true") {

        when {
            params {
                input         = "${projectDir}/tests/trimming/remove_adapter/samplesheets/input_with_adapters.csv"
                outdir        = "$outputDir"
                demultiplexer = 'bclconvert'
                skip_tools = 'samshee,falco' // Skip samshee validation which fails on our test samplesheets
                remove_adapter = true
            }
        }

        then {
            assertAll(
                { assert workflow.success },
                // Check that generated samplesheet has no adapter lines
                {
                    def samplesheetFiles = file("$outputDir").listFiles().findAll { it.name.endsWith('_no_adapters.csv') }
                    assert samplesheetFiles.size() > 0
                    def adapterlinesFound = false
                    samplesheetFiles.each { file ->
                        def content = file.text
                        if (content.contains('AdapterRead1') || content.contains('AdapterRead2') ||
                            content.contains('Adapter,AGATCGGAAGAG')) {
                            adapterlinesFound = true
                        }
                    }
                    assert !adapterlinesFound
                }
            )
        }
    }

    test("remove_adapter = false") {

        when {
            params {
                input         = "${projectDir}/tests/trimming/remove_adapter/samplesheets/input_with_adapters.csv"
                outdir        = "$outputDir"
                demultiplexer = 'bclconvert'
                skip_tools = 'samshee,falco' // Skip samshee validation which fails on our test samplesheets
                remove_adapter = false
            }
        }

        then {
            assertAll(
                { assert workflow.success }
                // When remove_adapter = false, we should NOT have *_no_adapters.csv files
            )
        }
    }

    test("remove_adapter = true with trim_fastq = true") {
        when {
            params {
                input         = "${projectDir}/tests/trimming/remove_adapter/samplesheets/input_with_adapters.csv"
                outdir        = "$outputDir"
                demultiplexer = 'bclconvert'
                skip_tools = 'samshee,falco' // Skip samshee validation which fails on our test samplesheets
                remove_adapter = true
                trim_fastq    = true
            }
        }

        then {
            assertAll(
                { assert workflow.success }
                // No need to check for fastp files in this test as it's already done in the first test
            )
        }
    }

    test("remove_adapter = true with trim_fastq = false") {
        when {
            params {
                input         = "${projectDir}/tests/trimming/remove_adapter/samplesheets/input_with_adapters.csv"
                outdir        = "$outputDir"
                demultiplexer = 'bclconvert'
                skip_tools = 'samshee,falco' // Skip samshee validation which fails on our test samplesheets
                remove_adapter = true
                trim_fastq    = false
            }
        }

        then {
            assertAll(
                { assert workflow.success }
                // No need to check for fastp files in this test as it's already done in the first test
            )
        }
    }

    // Additional test for when the samplesheet has no adapter lines
    test("remove_adapter = true with no adapters in samplesheet") {
        when {
            params {
                // Use the input samplesheet that points to no_adapters.csv
                input         = "${projectDir}/tests/trimming/remove_adapter/samplesheets/input_no_adapters.csv"
                outdir        = "$outputDir"
                demultiplexer = 'bclconvert'
                skip_tools = 'samshee,falco' // Skip samshee validation which fails on our test samplesheets
                remove_adapter = true
            }
        }

        then {
            assertAll(
                { assert workflow.success },
                // Check for the warning message in stdout or log
                {
                    // The warning can be in any of the log files
                    def logFiles = [
                        file("${launchDir}/.nextflow.log"),
                        file("${outputDir}/pipeline_info/execution_trace.txt"),
                        file("$outputDir/.nextflow.log")
                    ]

                    def warningFound = false
                    logFiles.each { logFile ->
                        if (logFile.exists()) {
                            def logContent = logFile.text
                            if (logContent.contains("no adapters were found in samplesheet")) {
                                warningFound = true
                            }
                        }
                    }

                    // Also check stdout captured by nf-test
                    if (workflow.stdout.contains("no adapters were found in samplesheet")) {
                        warningFound = true
                    }

                    assert warningFound, "Warning message about no adapters not found in logs"
                }
            )
        }
    }
}
