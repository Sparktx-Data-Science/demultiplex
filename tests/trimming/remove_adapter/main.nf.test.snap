{
    "remove_adapter = false": {
        "content": [
            0,
            {
                "Workflow": {
                    "nf-core/demultiplex": "v1.6.0dev"
                }
            },
            [
                "220422_M11111_0222_000000000-K9H97.csv",
                "pipeline_info",
                "pipeline_info/nf_core_pipeline_software_mqc_versions.yml"
            ],
            [
                "220422_M11111_0222_000000000-K9H97.csv:md5,2df2e405991814571c021dc8749c2a89"
            ]
        ],
        "meta": {
            "nf-test": "0.9.2",
            "nextflow": "25.01.0"
        },
        "timestamp": "2025-02-26T15:09:38.979909"
    },
    "remove_adapter = true": {
        "content": [
            0,
            {
                "Workflow": {
                    "nf-core/demultiplex": "v1.6.0dev"
                }
            },
            [
                "220422_M11111_0222_000000000-K9H97.lane1_no_adapters.csv",
                "pipeline_info",
                "pipeline_info/nf_core_pipeline_software_mqc_versions.yml"
            ],
            [
                "220422_M11111_0222_000000000-K9H97.lane1_no_adapters.csv:md5,ee5db2e12754e069998b0a96e535238c"
            ]
        ],
        "meta": {
            "nf-test": "0.9.2",
            "nextflow": "25.01.0"
        },
        "timestamp": "2025-02-26T15:09:32.699974"
    }
}